lib/collection/src/shard/collection_shard_distribution.rs

```diff
@@ -1,55 +1,66 @@
+}
+
 #[derive(Debug)]
 pub struct CollectionShardDistribution {
-    pub local: Vec<ShardId>,
-    pub remote: Vec<(ShardId, PeerId)>,
+    pub shards: HashMap<ShardId, ShardType>,
 }
 
 impl CollectionShardDistribution {
-    pub fn new(local: Vec<ShardId>, remote: Vec<(ShardId, PeerId)>) -> Self {
-        Self { local, remote }
-    }
-
     pub fn all_local(shard_number: Option<u32>) -> Self {
         Self {
             // This method is called only when distributed deployment is disabled
             // so if not specified it will suggest 1 shard per collection for better performance.
-            local: (0..shard_number.unwrap_or(1)).collect(),
-            remote: vec![],
+            shards: (0..shard_number.unwrap_or(1))
+                .map(|shard_id| (shard_id, ShardType::Local))
+                .collect(),
         }
     }
 
-    pub fn from_shard_to_peer(this_peer: PeerId, shard_to_peer: &HashMap<ShardId, PeerId>) -> Self {
-        let local = shard_to_peer
-            .iter()
-            .filter_map(|(shard, peer)| {
-                if peer == &this_peer {
-                    Some(*shard)
-                } else {
-                    None
-                }
-            })
-            .collect();
-
-        let remote = shard_to_peer
-            .iter()
-            .filter_map(|(&shard, &peer)| {
-                if peer != this_peer {
-                    Some((shard, peer))
-                } else {
-                    None
-                }
-            })
-            .clone()
-            .collect();
-
-        Self { local, remote }
+    pub fn from_shards_info(this_peer: PeerId, shards_info: HashMap<ShardId, ShardInfo>) -> Self {
+        Self {
+            shards: shards_info
+                .into_iter()
+                .map(|(shard_id, info)| match info {
+                    ShardInfo::ReplicaSet { replicas } => (
+                        shard_id,
+                        ShardType::ReplicaSet {
+                            local: replicas.contains_key(&this_peer),
+                            remote: replicas.into_keys().filter(|id| id != &this_peer).collect(),
+                        },
+                    ),
+                    ShardInfo::Single(peer_id) => {
+                        if peer_id == this_peer {
+                            (shard_id, ShardType::Local)
+                        } else {
+                            (shard_id, ShardType::Remote(peer_id))
+                        }
+                    }
+                })
+                .collect(),
+        }
     }
 
     pub fn shard_count(&self) -> usize {
-        self.local.len() + self.remote.len()
+        self.shards.len()
+    }
+
+    pub fn shard_replica_count(&self) -> usize {
+        self.shards
+            .values()
+            .map(|shard| match shard {
+                ShardType::ReplicaSet { local, remote } => *local as usize + remote.len(),
+                _ => 1,
```

Comment:
Nitpick a bit, I am unsure about the semantic here.
Is there such thing as a single replicas?